custom/icds_reports/sqldata/exports/dashboard_usage.py

```diff
@@ -0,0 +1,177 @@
+        filter_dict = {location_type + '_id': location_id, 'is_launched': 'yes'}
+        return AggAwc.objects.filter(**filter_dict).exists()
+
+    def get_excel_data(self):
+        excel_rows = []
+        filters = [['Generated at', india_now()]]
+        headers = ['Sr.No', 'State/UT Name', 'District Name', 'Block Name', 'Username', 'Level', 'Role',
+                   'Launched?', 'Last Login', 'Logged in the last week?', 'Total', 'Child', 'Pregnant Women',
+                   'Demographics', 'System Usage', 'AWC infrastructure', 'Child Growth Monitoring List',
+                   'ICDS - CAS Monthly Register', 'AWW Performance Report', 'LS Performance Report',
+                   'Take Home Ration']
+        excel_rows.append(headers)
+        serial_count = 0
+        for user_location in self.user.get_sql_locations(self.domain):
+            # getting the location types to retrieve for this user location
+            location_type_filter = {
+                self.get_location_id_string_from_location_type(user_location.location_type_name):
+                    user_location.get_id}
+            all_awc_locations = AwcLocation.objects.filter(**location_type_filter).values(*self.required_fields)
+            # converting the result set to matrix to fetch ancestors for a given location
+            location_matrix, location_ids =\
+                self.convert_rs_to_matrix(all_awc_locations, self.get_location_id_string_from_location_type(
+                    user_location.location_type_name))
+            users = self.get_users_by_location(location_ids)
+            records = list(ICDSAuditEntryRecord.objects.filter(url='/a/icds-dashboard-qa/cas_export')
+                           .annotate(indicator=KeyTextTransform('indicator', 'post_data')).values('indicator',
+                                                                                                  'username')
+                           .annotate(count=Count('indicator')).order_by('username', 'indicator'))
+            # accumulating the indicator counts
+            for user in users:
+                indicator_count = []
+                total_indicators = 0
+                for record in records:
+                    if record['username'] == user['username']:
+                        total_indicators += record['count']
+                        indicator_count.append(record['count'])
+
+                user_sql_location_ids = user['assigned_location_ids']
+                filter_dict = {}
+                if isinstance(user_sql_location_ids, str):
+                    filter_dict['location_id'] = user_sql_location_ids
+                else:
+                    filter_dict['location_id__in'] = user_sql_location_ids
+                user_sql_locations = SQLLocation.objects.filter(**filter_dict) \
+                    .values('location_id', 'location_type__name')
+                for user_sql_location in user_sql_locations:
```

Comment:
There are so many nested loops which will be very slow eventually and at higher level of hierarchy.

Reply:
The loops increase as there are more locations assigned to a single user which is not a general case i believe. The logic will cover national users as well